modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/richtext/RichTextArea.java

```diff
@@ -0,0 +1,2397 @@
+     */
+    public void applyStyle(TextPos start, TextPos end, StyleAttributeMap attrs) {
+        StyledTextModel m = getModel();
+        m.applyStyle(start, end, attrs, true);
+    }
+
+    /**
+     * When selection exists, deletes selected text.  Otherwise, deletes the character preceding the caret,
+     * possibly breaking up the grapheme clusters.
+     * This method does nothing if either control or the model is not editable, or the caret position is {@code null}.
+     * <p>
+     * This action can be changed by remapping the default behavior via {@link InputMap}.
+     *
+     * @see RichTextArea.Tags#BACKSPACE
+     */
+    public void backspace() {
+        execute(Tags.BACKSPACE);
+    }
+
+    /**
+     * Clears the document, creating an undo entry.
+     *
+     * @throws NullPointerException if the model is {@code null}
+     * @throws UnsupportedOperationException if the model is not {@link StyledTextModel#isWritable() writable}
+     */
+    public final void clear() {
+        TextPos end = getDocumentEnd();
+        replaceText(TextPos.ZERO, end, StyledInput.EMPTY, true);
+    }
+
+    /**
+     * Clears existing selection, if any.  This method is an alias for {@code getSelectionModel().clear()}.
+     */
+    public final void clearSelection() {
+        selectionModel.clear();
+    }
+
+    /**
+     * Clears the undo-redo stack of the underlying model.
+     * This method does nothing if the model is {@code null}.
+     */
+    public final void clearUndoRedo() {
+        StyledTextModel m = getModel();
+        if (m != null) {
+            m.clearUndoRedo();
+        }
+    }
+
+    /**
+     * When selection exists, copies the selected rich text to the clipboard in all the formats supported
+     * by the model.
+     * <p>
+     * This method does nothing if the model or the caret position is {@code null}.
+     * <p>
+     * This action can be changed by remapping the default behavior via {@link InputMap}.
+     * @see RichTextArea.Tags#COPY
+     */
+    public void copy() {
+        execute(Tags.COPY);
+    }
+
+    /**
+     * Copies the selected text in the specified format to the clipboard.
+     * This method does nothing if no selection exists or when the data format is not supported by the model.
+     *
+     * @param format the data format to use
+     */
+    public final void copy(DataFormat format) {
+        RichTextAreaSkin skin = richTextAreaSkin();
+        if (skin != null) {
+            skin.copyText(format);
+        }
+    }
+
+    /**
+     * Transfers the currently selected text to the clipboard,
+     * removing the current selection.
+     * <p>
+     * This method does nothing if the model or the caret position is {@code null}.
+     * The default implementation will attempt to copy the information first, then delete the selected text.
+     * Any failures, such as the control being not editable, or out of memory condition, will be silently ignored.
```

Comment:
I see what you are getting at, but I might recommend separating normal state conditions, such as "the control not being editable" from an error condition, such as OOM.

Btw, the implementation of `cut` currently throws an exception if not editable, so needs to be fixed.

Reply:
no, it does not - RichTextAreaBehavior.cut() L1232

```
                } catch(Exception | OutOfMemoryError e) {
                    RichUtils.provideErrorFeedback(control, e);
                }
```

one may question whether we should be handling OOME here and providing a feedback (which  currently dumps the stack trace to stderr which is not the best idea).

Reply:
To follow up on an offline discussion, all of the key-input-mapped methods that modify become no-ops if not editable and writable. So "cut" does nothing (i.e., does not do a copy) if the model is read-only or the control is not editable. This will be more consistent with TextArea and is easier to document.